webhookURLValidator.go

```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright 2021 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package ancla
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+)
+
+var (
+	errInvalidURL            = errors.New("invalid Config URL")
+	errInvalidFailureURL     = errors.New("invalid Failure URL")
+	errInvalidAlternativeURL = errors.New("invalid Alternative URL(s)")
+	errURLIsNotHTTPS         = errors.New("URL scheme is not HTTPS")
+	errInvalidHost           = errors.New("invalid host")
+	errIPGivenAsHost         = errors.New("cannot use IP as host")
+	errLoopbackGivenAsHost   = errors.New("cannot use loopback host")
+	errIPinInvalidSubnets    = errors.New("IP is within a blocked subnet")
+	errInvalidSubnet         = errors.New("invalid subnet")
+)
+
+// filterNil takes out all entries of Nil value from the slice.
+func filterNil(vs []ValidURLFunc) (filtered []ValidURLFunc) {
+	for _, v := range vs {
+		if v != nil {
+			filtered = append(filtered, v)
+		}
+	}
+	return
+}
+
+// GoodConfigURL parses the given webhook's Config.URL
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodConfigURL(vs []ValidURLFunc) ValidatorFunc {
+	vs = filterNil(vs)
+	return func(w Webhook) error {
+		parsedURL, err := url.ParseRequestURI(w.Config.URL)
+		if err != nil {
+			return fmt.Errorf("%w: %v", errInvalidURL, err)
+		}
+		for _, f := range vs {
+			err = f(parsedURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidURL, err)
+			}
+		}
+		return nil
+	}
+}
+
+// GoodFailureURL parses the given webhook's FailureURL
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodFailureURL(vs []ValidURLFunc) ValidatorFunc {
+	vs = filterNil(vs)
+	return func(w Webhook) error {
+		if w.FailureURL != "" {
```

Comment:
minor change: if you want to flatten this function, you could do something like:
```
if w.FailureURL == "" {
    return nil
}
...
```